File: NinePSharp/messages/Tgetlock.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Tgetlock : IMessage
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Tgetlock;
    public ushort Tag { get; }
    public uint Fid { get; }
    public byte LockType { get; }
    public ulong Start { get; }
    public ulong Length { get; }
    public uint ProcId { get; }
    public string ClientId { get; }

    public Tgetlock(uint size, ushort tag, uint fid, byte lockType, ulong start, ulong length, uint procId, string clientId)
    {
        Size = size;
        Tag = tag;
        Fid = fid;
        LockType = lockType;
        Start = start;
        Length = length;
        ProcId = procId;
        ClientId = clientId;
    }

    public Tgetlock(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        Fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        LockType = data[offset++];
        Start = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8)); offset += 8;
        Length = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8)); offset += 8;
        ProcId = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        ClientId = data.ReadString(ref offset);
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Tgetlock);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Fid); offset += 4;
        span[offset++] = LockType;
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), Start); offset += 8;
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), Length); offset += 8;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), ProcId); offset += 4;
        span.WriteString(ClientId, ref offset);
    }
}

File: NinePSharp.Messages.FSharp/messages/Tgetlock.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Tgetlock =
    { Size: uint32
      Tag: uint16
      Fid: uint32
      LockType: byte
      Start: uint64
      Length: uint64
      ProcId: uint32
      ClientId: string }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Tgetlock
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Tgetlock span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Fid); offset <- offset + 4
            span.[offset] <- this.LockType
            offset <- offset + 1
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.Start); offset <- offset + 8
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.Length); offset <- offset + 8
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.ProcId); offset <- offset + 4
            writeString this.ClientId span &offset

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let lockType = data.[offset]
        offset <- offset + 1
        let start = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        offset <- offset + 8
        let length = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        offset <- offset + 8
        let procId = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let clientId = readString data &offset
        { Size = size; Tag = tag; Fid = fid; LockType = lockType; Start = start; Length = length; ProcId = procId; ClientId = clientId }

